src/Acquisition/__init__.py

```diff
@@ -1,8 +1,833 @@
+from __future__ import absolute_import, print_function
+
+# pylint:disable=W0212,R0911,R0912
+
+import os
+import operator
+import sys
+import types
+
+import ExtensionClass
+
 from zope.interface import classImplements
 
-from _Acquisition import *
-from interfaces import IAcquirer
-from interfaces import IAcquisitionWrapper
+from .interfaces import IAcquirer
+from .interfaces import IAcquisitionWrapper
+
+class Acquired(object):
+    "Marker for explicit acquisition"
+
+
+_NOT_FOUND = object()  # marker
+
+###
+# Helper functions
+###
+
+
+def _has__of__(obj):
+    """Check whether an object has an __of__ method for returning itself
+    in the context of a container."""
+    # Note specifically this is a type check, not duck typing, or
+    # we get into cycles
+    return isinstance(obj, ExtensionClass.Base)
```

Comment:
The C implementation of has__of__ (actually a macro in ExtensionClass.h) checks not only that obj is an ExtensionClass instance, but also that its tp_descr_get slot filled (which for ExtensionClasses is equivalent to it having __of__ defined).

I'm running some Zope tests in PyPy using this branch of Acquisition, and hitting errors when I try to acquire something from an acquisition wrapper around a class that is an ExtensionClass but does not define __of__:

```
Traceback (most recent call last):
  File "/Users/davisagli/src/pypy-2.5.1-osx64/lib-python/2.7/unittest/case.py", line 320, in run
    self.setUp()
  File "/Users/davisagli/Code/Zope/develop/AccessControl/src/AccessControl/tests/testOwned.py", line 207, in setUp
    self.uf._doAddUser('user1', 'xxx', ['role1'], [])
  File "/Users/davisagli/Code/Zope/develop/AccessControl/src/AccessControl/userfolder.py", line 410, in _doAddUser
    self.data[name] = User(name, password, roles, domains)
  File "/Users/davisagli/Code/Zope/develop/Acquisition/src/Acquisition/__init__.py", line 327, in __getattribute__
    True, type(self)._IS_IMPLICIT, False, False)
  File "/Users/davisagli/Code/Zope/develop/Acquisition/src/Acquisition/__init__.py", line 262, in _Wrapper_findattr
    result = result.__of__(wrapper)
  File "/Users/davisagli/Code/Zope/eggs/persistent-4.0.8-py2.7.egg/persistent/persistence.py", line 238, in __getattribute__
    return _OGA(self, name)
AttributeError: 'PersistentMapping' object has no attribute '__of__'
```